protocol.md

```diff
@@ -0,0 +1,752 @@
+
+This frame continues a "call req" as described in the "fragmentation" section
+below.
+
+"flags" has the same definition as in "call req": to control fragmentation.
+
+## Payload: call res continue (type 0x14)
+
+Schema: `flags:1 csumtype:1 (csum:4){0,1} {continuation}`
+
+This frame continues a "call res" as described in "fragmentation" section
+below.
+
+"flags" has the same definition as in "call req": to control fragmentation.
+
+## Payload: claim (0xC1)
+
+Schema: `tracing:24`
+
+This message is used to claim or cancel a redundant request. When a request is
+sent to multiple nodes, as they start or finish work, depending on the option,
+they will tell the other nodes about this to minimize extra work. This claim
+message is sent from worker to to worker. The claimed request is referred to by
+its full zipkin tracing data, which was chosen by the originator of the first
+request.
+
+When a worker B receives a claim message from worker A for a tracing T that B
+doesn't know about, B will note this for a short time. If B later receives a
+request for T, B will silently ignore T. This is expected to be the common
+case, because the forwarding router will add some delay after sending to A and
+before sending to B.
+
+This is an implementation of Google's "The Tail at Scale" paper where they
+describe "backup requests with cross-server cancellation". Slides from a
+presentation about this paper are here:
+
+http://static.googleusercontent.com/media/research.google.com/en/us/people/jeff/Berkeley-Latency-Mar2012.pdf
+
+The relevant section starts on page 39. A video of this talk is here:
+
+http://youtu.be/C_PxVdQmfpk?t=26m45s
+
+## Payload: ping req (0xD0)
+
+Used to verify that the protocol is functioning correctly over a connection.
+The receiving side will send back a "ping res" message, but this is not
+expected to be visible to the application. If more detailed health checking and
+validation checks are desired, these can be implemented at a higher level with
+"call req" and "call res" messages.
+
+This message type has no body.
+
+## Payload: ping res (0xD1)
+
+Always sent in response to a "ping req". Sending this does not necessarily mean
+the service is "healthy." It only validates connectivity and basic protocol
+functionality.
+
+This message type has no body.
+
+## Payload: error (0xFF)
+
+Schema: `code:1 id:4 message~2`
+
+Response message to a failure at the protocol level or when the system was
+unable to invoke the requested RPC for some reason. Application errors do not
+go here. Application errors are sent with "call res" messages and application
+specific exception data in the args.
+
+### code:1
+
+code | name
+-----|----------------------
+0x00 | invalid
+0x01 | timeout
+0x02 | cancelled
+0x03 | busy
+0x04 | declined
+0x05 | unexpected error
+0x06 | bad request
+0xFF | fatal protocol error
+
+#### Error code `0x00` -- invalid
+
+Not a valid value for `code`. Do not use.
```

Comment:
Personally, I would move the descriptions into the table instead of adding separate sections for them another level deeper.